### PR TITLE
🌟 Nova: Associative Retrieval ("Fishing") Module

### PR DESCRIPTION
**Pull request by madmax983**
This PR introduces "Associative Retrieval" (codenamed "Fishing"), a new experimental feature that enables multi-modal search.

**The Problem:**
Vector search finds semantically similar things. Graph traversal finds structurally related things. But human memory is *associative* - it links concepts based on meaning, connection, AND recency.

**The Solution:**
The `FishingRod` API allows users to:
1.  **Cast a Line:** Start with a node or a raw vector (the "Bait").
2.  **Find the School:** Use vector search to find initial candidates.
3.  **Spread the Net:** Expand to direct neighbors in the graph.
4.  **Weigh the Catch:** Score results based on a weighted combination of:
    -   Vector Similarity
    -   Graph Connection (Bonus for neighbors)
    -   Freshness (Decay function based on `commit_timestamp`)

**Usage:**
```rust
let rod = FishingRod::new(&db);
let bait = Bait::Vector(vec![0.1, 0.2, ...]);
let trip = FishingTrip {
    limit: 10,
    depth: 1,
    vector_weight: 1.0,
    graph_weight: 0.5,
    freshness_weight: 0.2,
};
let catches = rod.cast(bait, trip)?;
```

This feature is isolated in `src/experimental/fishing.rs` and gated by the `nova` feature flag.

---
*PR created automatically by Jules for task [11022217858181333547](https://jules.google.com/task/11022217858181333547) started by @madmax983*

### CODE REVIEW

**Comment by madmax983** [comment]
> ## Code Review: Associative Retrieval ("Fishing") Module
> Thank you for this interesting experimental feature! The concept of combining vector similarity, graph proximity, and temporal freshness is compelling for associative memory patterns. Here's my detailed review:
> 
> ## 🟢 Strengths
> ### 1. **Clear Conceptual Design**
> The "fishing" metaphor is well-chosen and consistently applied throughout the code. The documentation clearly explains the approach and use cases.
> 
> ### 2. **Appropriate Isolation**
> * Correctly gated behind the `nova` feature flag
> * Placed in `experimental/` module where it belongs
> * Doesn't modify core APIs
> 
> ### 3. **Good API Ergonomics**
> * `FishingTrip` config with sensible defaults
> * `Bait` enum allows starting from either a node or raw vector
> * `Catch` struct with provenance tracking is excellent for explainability
> 
> ### 4. **Test Coverage**
> The workflow test demonstrates the core functionality and validates expected behavior.
> 
> ## 🟡 Issues Found
> ### **CRITICAL: Potential Panic in Production Code**
> **Location:** `fishing.rs:180`
> 
> ```rust
> catches.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal));
> ```
> 
> **Issue:** While this code correctly handles NaN with `unwrap_or`, the pattern is inconsistent with the project's strict "no unwrap in production" policy from [CODING_STANDARDS.md](https://github.com/madmax983/GallifreyDB/blob/trunk/docs/CODING_STANDARDS.md).
> 
> **Recommendation:** This is actually safe because `partial_cmp` only returns `None` for NaN, which is handled. However, for consistency with codebase standards, consider adding a comment:
> 
> ```rust
> // SAFETY: partial_cmp only returns None for NaN, which we handle with unwrap_or
> catches.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal));
> ```
> 
> Or use the more explicit pattern from `semantic_pathfinding.rs`:
> 
> ```rust
> catches.sort_by(|a, b| {
>     b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal)
> });
> ```
> 
> ### **MAJOR: Silent Failure When No Vector Indexes Exist**
> **Location:** `fishing.rs:97-105`
> 
> ```rust
> let indexes = self.db.list_vector_indexes();
> if let Some(idx_info) = indexes.first() {
>     self.db.search_vectors_in(&idx_info.property_name, embedding, config.limit)?
> } else {
>     return Ok(vec\![]); // No vector indexes, no fish.
> }
> ```
> 
> **Issue:** The function silently returns empty results when no vector index exists. This makes it difficult to distinguish between "no matches found" and "vector search is not configured."
> 
> **Recommendation:** Return a proper error:
> 
> ```rust
> } else {
>     return Err(crate::utils::error::Error::Vector(
>         crate::utils::error::VectorError::IndexError(
>             "No vector indexes configured. Call enable_vector_index() first.".to_string()
>         )
>     ));
> }
> ```
> 
> ### **MODERATE: Magic Property Name Selection**
> **Location:** `fishing.rs:99-100`
> 
> ```rust
> if let Some(idx_info) = indexes.first() {
>     self.db.search_vectors_in(&idx_info.property_name, embedding, config.limit)?
> ```
> 
> **Issue:** When multiple vector indexes exist, the code arbitrarily picks the first one. This is unpredictable and could lead to confusing results.
> 
> **Recommendation:**
> 
> 1. Add a `property_name: Option<String>` field to `FishingTrip` to allow explicit selection
> 2. If `None`, pick the first available (current behavior) but document this clearly
> 3. Consider warning in docs about multi-property setups
> 
> ```rust
> pub struct FishingTrip {
>     pub limit: usize,
>     pub depth: usize,
>     pub vector_weight: f32,
>     pub graph_weight: f32,
>     pub freshness_weight: f32,
>     /// Optional vector property to search. If None, uses the first available index.
>     /// Required when multiple vector indexes are configured.
>     pub vector_property: Option<String>,
> }
> ```
> 
> ### **MINOR: Inconsistent Error Handling in `Bait::Node` Path**
> **Location:** `fishing.rs:84-95`
> 
> ```rust
> Bait::Node(node_id) => {
>     // Comments mention uncertainty about which property to use
>     // Relies on find_similar implementation details
>     self.db.find_similar(node_id, config.limit)?
> }
> ```
> 
> **Issue:** The comment reveals uncertainty about how this works. Looking at the code, `find_similar` may fail if the node doesn't have the expected vector property.
> 
> **Recommendation:** Either:
> 
> 1. Document the expected behavior clearly ("Uses the node's embedding from the first available vector index")
> 2. Or allow `Bait::Node` to specify which property to use:
> 
> ```rust
> pub enum Bait {
>     /// Start with an existing node's embedding from the specified property.
>     Node { id: NodeId, property: Option<String> },
>     /// Start with a raw embedding vector.
>     Vector(Vec<f32>),
> }
> ```
> 
> ### **MINOR: Potential Performance Issue**
> **Location:** `fishing.rs:150-165` (freshness calculation)
> 
> **Issue:** For each candidate node, the code calls `db.get_node()` which may be expensive if there are many candidates. This happens even when `freshness_weight == 0.0`.
> 
> **Recommendation:**
> 
> ```rust
> // Early exit if freshness is not weighted
> if config.freshness_weight > 0.0 {
>     let now_micros = time::now().wallclock();
>     // ... rest of freshness logic
> }
> ```
> 
> ### **MINOR: Missing Edge Labels Filter**
> **Location:** `fishing.rs:125-131` (graph traversal)
> 
> **Issue:** The graph expansion follows ALL outgoing edges regardless of edge type. For real-world graphs, you likely want to filter by edge label (e.g., only follow "RELATED_TO" edges, not "CREATED_BY").
> 
> **Recommendation:** Add `edge_labels: Option<Vec<String>>` to `FishingTrip`:
> 
> ```rust
> pub struct FishingTrip {
>     // ... existing fields
>     /// Optional edge labels to follow during graph expansion. If None, follows all edges.
>     pub edge_labels: Option<Vec<String>>,
> }
> ```
> 
> ## 🔵 Suggestions for Future Enhancements
> 1. **Bi-directional Traversal:** Currently only follows outgoing edges. Consider allowing incoming edges or both.
> 2. **Depth > 1:** The code mentions "Currently only depth 1 is supported" but doesn't validate this. Add validation:
> 
> ```rust
> if config.depth > 1 {
>     return Err("Multi-hop traversal not yet implemented".into());
> }
> ```
> 
> 3. **Vector Search Depth:** Consider separating the initial vector search `k` from the final result `limit`. You might want to search for 100 similar nodes but only return top 10 after graph expansion.
> 4. **Deduplication:** The code correctly avoids duplicating scores for nodes found via multiple paths, but the score accumulation is simple addition. Consider more sophisticated aggregation (max, weighted average, etc.).
> 5. **Provenance Detail:** The provenance string is helpful but could be richer. Consider tracking:
>    
>    * Vector similarity score
>    * Which edge(s) connected the node
>    * Freshness contribution
> 
> ## 🔴 Missing: Pre-Commit Checks
> **BLOCKING ISSUE:** According to [CLAUDE.md](https://github.com/madmax983/GallifreyDB/blob/trunk/CLAUDE.md#-mandatory-pre-commit-quality-checks), you MUST run these before every commit:
> 
> 1. `cargo clippy --all-targets --all-features -- -D warnings`
> 2. `cargo fmt --all`
> 3. `cargo test`
> 
> **Action Required:** Please verify these pass. The PR description doesn't mention running these checks.
> 
> ## 📊 Test Coverage Analysis
> ### Current Coverage
> The test covers:
> 
> * ✅ Basic workflow (vector search → graph expansion)
> * ✅ Vector bait path
> * ✅ Provenance tracking
> 
> ### Missing Coverage
> Consider adding tests for:
> 
> * ❌ `Bait::Node` variant
> * ❌ No vector index configured (should error with my recommendation)
> * ❌ Multiple vector indexes (which one gets picked?)
> * ❌ Freshness weighting (currently disabled with `freshness_weight: 0.0`)
> * ❌ Empty results (no similar nodes found)
> * ❌ Depth = 0 (should only do vector search)
> * ❌ Edge cases: NaN scores, zero weights, limit = 0
> 
> ## ✅ Summary
> **Verdict:** This is a solid experimental feature with a clear use case. The code is well-structured and documented. However, it needs refinement before merging:
> 
> ### Required Changes (Before Merge):
> 1. ✅ Fix silent failure when no vector indexes exist (return error)
> 2. ✅ Add `property_name` to `FishingTrip` for multi-index scenarios
> 3. ✅ Verify pre-commit checks pass
> 4. ✅ Add comment explaining `unwrap_or` safety in sort
> 
> ### Recommended Changes:
> 1. Add edge label filtering
> 2. Validate `depth > 1` and return error
> 3. Optimize freshness check with early exit
> 4. Expand test coverage
> 
> ### Future Enhancements:
> 1. Bi-directional traversal
> 2. Separate vector search `k` from final `limit`
> 3. More sophisticated score aggregation
> 4. Richer provenance tracking
> 
> Great work on this feature! The associative retrieval concept aligns perfectly with GallifreyDB's LLM integration goals. Once the critical issues are addressed, this will be a valuable addition to the experimental module.
> 
> Let me know if you'd like help implementing any of these suggestions!

@jules

File: src/experimental/fishing.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+/// The input to the fishing algorithm.
+#[derive(Debug, Clone)]
+pub enum Bait {
+    /// Start with an existing node in the graph.
+    Node(NodeId),
+    /// Start with a raw embedding vector.
+    Vector(Vec<f32>),
+}
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/experimental/fishing.rs:73]: > high
> 
> The `Bait::Vector(Vec<f32>)` variant is ambiguous when multiple vector indexes are present. The current implementation in `cast` arbitrarily picks the first available index, which can lead to non-deterministic and unexpected behavior. To make this explicit and robust, I suggest changing the `Bait` enum to include the property name for the vector search. This will also require updating the `cast` function and the test case in `test_fishing_workflow`.

@jules

File: src/experimental/fishing.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+                // We need to know which property to search.
+                // Ideally `Bait` should specify property, but for "magic" we can try to guess or search all.
+                // For this MVP, let's look for the first enabled vector index.
+                let indexes = self.db.list_vector_indexes();
+                if let Some(idx_info) = indexes.first() {
+                    self.db.search_vectors_in(&idx_info.property_name, embedding, config.limit)?
+                } else {
+                    return Ok(vec![]); // No vector indexes, no fish.
+                }
+            }
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/experimental/fishing.rs]: > high
> 
> Following the recommended change to the `Bait` enum, this `match` arm should be updated to use the explicit `property_name` provided in the `Bait`. This removes the fragile logic of picking the first available index and makes the vector search behavior explicit and reliable.
> 
> ```rust
>             Bait::Vector { property_name, ref embedding } => {
>                 // Use the explicit property name for the search.
>                 self.db.search_vectors_in(property_name, embedding, config.limit)?
>             }
> ```

@jules

File: src/experimental/fishing.rs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+            for (target, source) in neighbors {
+                let current_score = *candidate_scores.get(&target).unwrap_or(&0.0);
+                // Simple additive boost for being a neighbor
+                let new_score = current_score + config.graph_weight;
+                candidate_scores.insert(target, new_score);
+
+                provenance.entry(target)
+                    .and_modify(|p| *p += &format!("\nLinked from Node {}", source))
+                    .or_insert_with(|| format!("Linked from Node {}", source));
+            }
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/experimental/fishing.rs:181]: > medium
> 
> The graph traversal logic currently collects all neighbors into an intermediate `neighbors` vector before processing them. This can be inefficient for large result sets from the vector search, as it requires an extra allocation and iteration. You can improve performance and reduce memory usage by processing neighbors as they are discovered within a single loop.
> 
> ```rust
>             for (source_node, _) in school.iter() {
>                 let edges = self.db.get_outgoing_edges(*source_node);
>                 for edge_id in edges {
>                     if let Ok(target) = self.db.get_edge_target(edge_id) {
>                         // Simple additive boost for being a neighbor
>                         *candidate_scores.entry(target).or_insert(0.0) += config.graph_weight;
> 
>                         provenance.entry(target)
>                             .and_modify(|p| *p += &format!("\nLinked from Node {}", source_node))
>                             .or_insert_with(|| format!("Linked from Node {}", source_node));
>                     }
>                 }
>             }
> ```

@jules

File: src/experimental/fishing.rs [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+                        // Decay function: 1 / (1 + age in hours)
+                        let age_hours = age_seconds / 3600.0;
+                        let freshness_score = 1.0 / (1.0 + age_hours);
+
+                        candidate_scores.entry(node_id).and_modify(|s| {
+                            *s += freshness_score * config.freshness_weight;
+                        });
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/experimental/fishing.rs:213]: > medium
> 
> The freshness calculation involves iterating over all candidate nodes and calling `self.db.get_node(node_id)` for each one. This creates an N+1 query problem, which can be a significant performance bottleneck if there are many candidates. While `get_node` is fast, the overhead of many individual calls can add up. Consider adding a batch `get_nodes` method to `GallifreyDB` to fetch all candidate nodes in a single operation. This would reduce overhead and improve performance for this step.

@jules

File: src/experimental/fishing.rs [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        };
+
+        let catches = rod.cast(bait, trip).unwrap();
+
+        // We expect:
+        // 1. N1 (Perfect match or close)
+        // 2. N2 (Close match)
+        // 3. N3 (Linked from N2)
+
+        assert!(catches.len() >= 2);
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/experimental/fishing.rs:303]: > medium
> 
> The assertion `assert!(catches.len() >= 2)` is quite loose. The test setup creates three nodes that should be returned: `n1` (from `props1`, an exact vector match), `n2` (similar vector), and `n3` (neighbor of `n2`). The test should be more specific and assert that all three are present in the results. Currently, the test doesn't check for `n1` at all. Consider strengthening the assertions to verify all expected results.
> 
> ```rust
>         assert_eq!(catches.len(), 3, "Expected to find n1 (exact match), n2 (similar), and n3 (neighbor)");
> ```

@jules

File: src/experimental/fishing.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (madmax983) [reply src/experimental/fishing.rs]: > In the `Bait::Vector` branch you're selecting the index with `list_vector_indexes().first()`, but elsewhere in the codebase the "default" vector index is determined via `get_default_vector_property_name()` (alphabetical min) and exposed through `find_similar_by_embedding`/`prepare_vector_search_raw` (see `src/storage/current/mod.rs:983-997` and `1513-1519`). This means associative retrieval with a raw vector can use a different index than node-based retrieval in multi-index setups, which is inconsistent with the established default-selection convention. To keep behavior predictable and aligned with the rest of the API, consider delegating to `GallifreyDB::find_similar_by_embedding` (or otherwise reusing the same default-index logic) instead of picking the first entry from `list_vector_indexes()`.

@jules

File: src/experimental/fishing.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (madmax983) [reply src/experimental/fishing.rs:213]: > The freshness-scoring branch (`if config.freshness_weight > 0.0 { ... }`) and the `Bait::Node` path in `cast` are not exercised by `test_fishing_workflow`, so regressions in timestamp handling or node-based bait behavior wouldn't be caught by tests. Given that other experimental modules (e.g., `semantic_pathfinding`) have targeted tests for their key code paths, it would be good to add at least one test that sets a non-zero `freshness_weight` and one that uses `Bait::Node` to verify the combined scoring and provenance behavior.

@jules
